contracts/GuardiansWallet.sol

```diff
@@ -0,0 +1,195 @@
+            balance.fees += amount48;
+            vars.totalFees = vars.totalFees.add(amount);
+
+            amount48 = toUint48Granularity(stakingRewards[i]);
+            amount = toUint256Granularity(amount48);
+            balance.stakingRewards += amount48;
+            vars.totalStakingRewards = vars.totalStakingRewards.add(amount);
+
+            balances[guardians[i]] = balance;
+        }
+
+        feesToken.transferFrom(msg.sender, address(this), vars.totalFees);
+        bootstrapToken.transferFrom(msg.sender, address(this), vars.totalBootstrapRewards);
+        stakingToken.transferFrom(msg.sender,  address(this), vars.totalStakingRewards);
+
+        emit RewardsAssigned(guardians, stakingRewards, fees, bootstrapRewards);
+    }
+
+    function setMaxDelegatorsStakingRewards(uint32 _maxDelegatorsStakingRewardsPercentMille) external onlyFunctionalOwner {
+        require(_maxDelegatorsStakingRewardsPercentMille <= 100000, "_maxDelegatorsStakingRewardsPercentMille must not be larger than 100000");
+        maxDelegatorsStakingRewardsPercentMille = _maxDelegatorsStakingRewardsPercentMille;
+        emit MaxDelegatorsStakingRewardsChanged(_maxDelegatorsStakingRewardsPercentMille);
+    }
+
+    function getBootstrapBalance(address addr) external view returns (uint256) {
+        return toUint256Granularity(balances[addr].bootstrapRewards);
+    }
+
+    function withdrawBootstrapFunds() external {
+        address guardianAddress = getGuardiansRegistrationContract().resolveGuardianAddress(msg.sender);
+
+        uint48 amount = balances[guardianAddress].bootstrapRewards;
+
+        balances[guardianAddress].bootstrapRewards = 0;
+
+        emit BootstrapRewardsWithdrawn(guardianAddress, toUint256Granularity(amount));
+        require(transfer(bootstrapToken, guardianAddress, amount), "GuardiansWallet::withdrawBootstrapFunds - insufficient funds");
+    }
+
+    function getStakingRewardBalance(address addr) external view returns (uint256) {
+        return toUint256Granularity(balances[addr].stakingRewards);
+    }
+
+    struct DistributorBatchState {
+        uint256 fromBlock;
+        uint256 toBlock;
+        uint256 nextTxIndex;
+        uint split;
+    }
+    mapping (address => DistributorBatchState) distributorBatchState;
+
+    function isDelegatorRewardsBelowThreshold(uint256 delegatorRewards, uint256 totalRewards) private view returns (bool) {
+        return delegatorRewards.mul(100000) <= uint(maxDelegatorsStakingRewardsPercentMille).mul(totalRewards.add(toUint256Granularity(1))); // +1 is added to account for rounding errors
+    }
+
+    struct DistributeOrbsTokenStakingRewardsVars {
+        bool firstTxBySender;
+        address guardianAddr;
+        uint256 delegatorsAmount;
+    }
+    function distributeStakingRewards(uint256 totalAmount, uint256 fromBlock, uint256 toBlock, uint split, uint txIndex, address[] calldata to, uint256[] calldata amounts) external {
```

Comment:
Carefully review the removal of onlyWhenActive.
Do not call getDelegationsContract().refreshStakeNotification(vars.guardianAddr) if not active

Reply:
May still distribute to staking wallet it accepts new staking unless set otherwise

Reply:
Anyone can call `refreshStakeNotification`, if it's a problem for `delegations` then it should be locked anyway. I like the all-or-nothing approach about locking functions.